src/v/cluster/distributed_kv_stm.h

```diff
@@ -0,0 +1,356 @@
+
+        auto res = co_await replicate_and_wait(
+          make_coordinator_assignment_batch(
+            key, result, coordinator_assignment_status::assigned));
+
+        if (res != errc::success) {
+            co_return res;
+        }
+        co_return co_await coordinator(key);
+    }
+
+    ss::future<result<std::optional<Value>>> get(Key key) {
+        auto holder = _gate.hold();
+        auto units = co_await _snapshot_lock.hold_read_lock();
+        if (!co_await sync(sync_timeout)) {
+            co_return errc::not_leader;
+        }
+        auto it = _kvs.find(key);
+        if (it == _kvs.end()) {
+            co_return std::nullopt;
+        }
+        co_return it->second;
+    }
+
+    ss::future<errc> put(absl::flat_hash_map<Key, Value> kvs) {
+        auto holder = _gate.hold();
+        auto units = co_await _snapshot_lock.hold_read_lock();
+        if (!co_await sync(sync_timeout)) {
+            co_return errc::not_leader;
+        }
+        co_return co_await replicate_and_wait(
+          make_kv_data_batch(std::move(kvs)));
+    }
+
+    ss::future<errc> remove(Key key) {
+        auto holder = _gate.hold();
+        auto units = co_await _snapshot_lock.hold_read_lock();
+        auto it = _kvs.find(key);
+        if (it == _kvs.end()) {
+            co_return errc::success;
+        }
+        co_return co_await replicate_and_wait(
+          make_kv_data_batch_remove_key<Key, Value>(key));
+    }
+
+    ss::future<result<size_t>> repartition(size_t new_partition_count) {
+        auto holder = _gate.hold();
+        if (!_is_routing_partition) {
+            co_return errc::invalid_request;
+        }
+        if (!co_await sync(sync_timeout)) {
+            co_return errc::not_leader;
+        }
+        auto units = co_await _snapshot_lock.hold_read_lock();
+        auto repartition_units = co_await _repartitioning_lock.get_units();
+        if (_num_partitions && _num_partitions.value() == new_partition_count) {
+            co_return _num_partitions.value();
+        }
+        if (_num_partitions && _num_partitions.value() > new_partition_count) {
+            // reducing the number of partitions is disallowed without
+            // a proper migration protocol.
+            co_return errc::invalid_request;
+        }
+        auto res = co_await replicate_and_wait(
+          make_repartitioning_batch(new_partition_count));
+
+        if (res != errc::success) {
+            co_return res;
+        }
+        vassert(
+          _num_partitions.has_value(),
+          "unexpected state, stm has not applied {}",
+          new_partition_count);
+        co_return _num_partitions.value();
+    }
+
+private:
+    static constexpr model::partition_id routing_partition{0};
+    static constexpr std::chrono::seconds sync_timeout{5};
+    // a guardrail until an inactivity based expiration is
+    // implemented.
+    static constexpr size_t keys_limit = 10000;
```

Comment:
FWIW the coordinator doesn't have a limit, so something to keep in mind if we ever want to dynamically create more partitions when more transforms are deployed/scaled up. No action is needed now, I just wanted to point that out.

One idea is maybe we could do range based sharding instead of hashed, then we can compress the coordinator data by just storing ranges (with the downside that it's easier to hotspot partitions with range sharding).

Reply:
yep, I think swapping out hash with range partitioning should be a fairly localized and simple change once we know what scale we are looking for.